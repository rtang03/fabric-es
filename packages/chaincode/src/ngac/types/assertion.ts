export interface Assertion {
  sid: string;
  assertion: boolean;
  message?: string;
}
