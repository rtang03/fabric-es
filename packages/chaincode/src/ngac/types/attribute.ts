export interface Attribute {
  type: '1' | 'N';
  key: string;
  value: string | string[];
  alias?: string;
  disabled?: boolean;
  immutable?: boolean;
}
