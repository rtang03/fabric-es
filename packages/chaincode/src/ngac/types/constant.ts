export enum CONTEXT {
  ISSUER_ORGNAME = 'issuer_orgname',
  ISSUER_CN = 'issuer_cn',
  SUBJECT_OUNAME = 'subject_ouname',
  SUBJECT_ORGNAME = 'subject_orgname',
  SUBJECT_CN = 'subject_cn',
  INVOKER_ID = 'invoker_id',
  INVOKER_MSPID = 'invoker_mspid'
}

export enum RESOURCE {
  ENTITYNAME = 'entityName',
  VERSION = 'version',
  CREATOR_MSPID = 'creator_mspid',
  CREATOR_CN = 'creator_cn',
  CREATOR_ID = 'creator_id'
}

export enum NAMESPACE {
  MODEL = 'model',
  ORG = 'org',
  ENTITY = 'entity',
  MSP_ATTRIBUTE = 'mspattr',
  ENTITY_ATTRIBUTE = 'entityattr',
  RESOURCE_ATTRIBUTE = 'resattr',
  POLICY = 'policy'
}
