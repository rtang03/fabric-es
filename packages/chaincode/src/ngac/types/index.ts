export * from './resource';
export * from './action';
export * from './policy';
export * from './attribute';
export * from './constant';
export * from './ngacRepo';
export * from './assertion';
