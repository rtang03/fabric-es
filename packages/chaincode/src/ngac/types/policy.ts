export type Effect = 'Allow' | 'Deny';

export interface Policy {
  key: string;
  policyClass?: string;
  sid: string;
  allowedEvents: string[];
  attributes: {
    uri: string;
    // creatorCN?: string;
    // creatorID?: string;
    // creatorMSPID?: string;
    // invokerID?: string;
    // invokerMSPID?: string;
    // invokerSubjectCN?: string;
    // invokerIssuerCN?: string;
    // type?: string;
    // entityName?: string;
    // version?: string;
  };
  condition?: {
    hasList?: Record<string, string>;
    stringEquals?: Record<string, string>;
  };
  effect: string;
}
