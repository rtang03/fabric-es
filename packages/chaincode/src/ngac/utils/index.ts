export * from './makeKey';
export * from './createResource';
export * from './createPolicy';
export * from './createId';
export * from './hasList';
export * from './stringEquals';
