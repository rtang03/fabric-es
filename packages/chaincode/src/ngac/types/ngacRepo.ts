import { Attribute } from './attribute';
import { Policy } from './policy';
import { Resource } from './resource';

export interface NgacRepo {
  // add one policy
  addPolicy: (policy: Policy) => Promise<Policy>;
  // add and replace one attribute group to a MSPID
  addMSPAttr: (resource: Resource) => Promise<Attribute[]>;
  // add and replace one attribute group to a reource URI
  addResourceAttr: (resource: Resource) => Promise<Attribute[]>;
  deleteMSPAttrByMSPID: (mspid: string) => Promise<string>;
  // delete all policies by id
  deletePolicyById: (id: string) => Promise<string[]>;
  // delete one policy by id
  deletePolicyByIdSid: (id: string, sid: string) => Promise<string>;
  // delete attribute group by URI
  deleteReourceAttrByURI: (uri: string) => Promise<string>;
  // delete attribute group by MSPID
  getMSPAttrByMSPID: (mspid: string) => Promise<Attribute[]>;
  // get policies by id
  getPolicyById: (id: string) => Promise<Policy[]>;
  // get one policy
  getPolicyByIdSid: (id: string, sid: string) => Promise<Policy>;
  // get array of attribute group by URI path segment
  // this API may be useless; no such use case. This is not exposed as chaincode interface
  getResourceAttrGroupByURI: (uri: string) => Promise<Attribute[][]>;
  // get attribte gropu by URI
  getResourceAttrByURI: (uri: string) => Promise<Attribute[]>;
  // update and insert the attribute group, by appending the existing one
  upsertResourceAttr: (resource: Resource) => Promise<Attribute[]>;
}
