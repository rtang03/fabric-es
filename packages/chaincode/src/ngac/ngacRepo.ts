import { Context } from 'fabric-contract-api';
import { assign, values } from 'lodash';
import stateList from './ledger-api/statelist';
import { Attribute, NAMESPACE as NS, NgacRepo, Policy } from './types';
import { makeKey } from './utils';

export const ngacRepo: (context: Context) => NgacRepo = context => ({
  addMSPAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).addState(
          resource.key,
          resource.mspAttrs
        )
      : null,
  addResourceAttr: async resource =>
    resource
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resource.resourceAttrs
        )
      : null,
  upsertResourceAttr: async resource => {
    if (!resource) return null;
    const attributes = await stateList<Attribute[]>(
      NS.RESOURCE_ATTRIBUTE,
      context
    ).getState(resource.key);

    const newAttrs = {};
    resource.resourceAttrs.forEach(attr => (newAttrs[attr.key] = attr));
    const existingAttrs = {};
    attributes.forEach(attr => (existingAttrs[attr.key] = attr));
    const resultAttrs = values<Attribute>(assign({}, existingAttrs, newAttrs));

    return resultAttrs
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).addState(
          resource.key,
          resultAttrs.map(attr => {
            if (attr.key === 'version') {
              let version = parseInt(attr.value as string, 10);
              return {
                type: '1',
                key: 'version',
                value: `${++version}`,
                immutable: false
              };
            }
            return attr;
          })
        )
      : null;
  },
  addPolicy: async policy =>
    policy
      ? stateList<Policy>(NS.POLICY, context).addState(policy.key, policy)
      : null,
  deleteMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList(NS.MSP_ATTRIBUTE, context).deleteStateByKey(makeKey([mspid]))
      : null,
  deletePolicyById: async id =>
    id
      ? stateList(NS.POLICY, context).deleteStatesByKeyRange([
          JSON.stringify(id)
        ])
      : null,
  deletePolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList(NS.POLICY, context).deleteStateByKey(makeKey([id, sid]))
      : null,
  deleteReourceAttrByURI: async uri =>
    uri
      ? stateList(NS.RESOURCE_ATTRIBUTE, context).deleteStateByKey(
          makeKey(uri.split('/'))
        )
      : null,
  getMSPAttrByMSPID: async mspid =>
    mspid
      ? stateList<Attribute[]>(NS.MSP_ATTRIBUTE, context).getState(
          makeKey([mspid])
        )
      : null,
  getPolicyById: async id =>
    id
      ? stateList<Policy>(NS.POLICY, context).getQueryResult([
          JSON.stringify(id)
        ])
      : null,
  getPolicyByIdSid: async (id, sid) =>
    id && sid
      ? stateList<Policy>(NS.POLICY, context).getState(makeKey([id, sid]))
      : null,
  getResourceAttrGroupByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getQueryResult(
          uri.split('/').map(part => JSON.stringify(part))
        )
      : null,
  getResourceAttrByURI: async uri =>
    uri
      ? stateList<Attribute[]>(NS.RESOURCE_ATTRIBUTE, context).getState(
          makeKey(uri.split('/'))
        )
      : null
});
