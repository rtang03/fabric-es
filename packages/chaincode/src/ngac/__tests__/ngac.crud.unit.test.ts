import { ngacRepo } from '../ngacRepo';
import { NAMESPACE, Policy } from '../types';
import { createId, createPolicy, createResource } from '../utils';
import { createMSPResource } from '../utils/createMSPResource';
import { mspAttributeDb, policyDb, resourceAttributeDb } from './__utils__';

jest.mock('../ledger-api/statelist');

const context: any = {
  clientIdentity: {
    getMSPID: jest.fn(),
    getID: jest.fn(),
    getX509Certificate: jest.fn()
  },
  resourceAttributeDb,
  mspAttributeDb,
  policyDb
};

const x509id = createId(['Org1MSP', 'Admin@example.com']);
// 'x509::/O=Dev/OU=client/CN=Admin@example.com::/O=Dev/OU=Dev/CN=rca';
const sid = 'allowCreateDocument';
const mspid = 'Org1MSP';
const uri = 'model/Org1MSP/dev_ngac_example1';
context.clientIdentity.getMSPID.mockImplementation(() => 'Org1MSP');
context.clientIdentity.getX509Certificate.mockImplementation(() => ({
  subject: { commonName: 'Admin@example.com' },
  issuer: { commonName: 'rca-org1' }
}));

describe('NgacRepo CRUD Tests', () => {
  beforeEach(() => {
    context.clientIdentity.getID.mockImplementation(() => x509id);
  });

  it('should getPolicyById by x509id', async () =>
    ngacRepo(context)
      .getPolicyById(x509id)
      .then(policies => policies.forEach(({ policyClass }) => expect(policyClass).toBe('event-creation'))));

  it('should fail getPolicyById: wrong x509id', async () =>
    ngacRepo(context)
      .getPolicyById('incorrect x509id')
      .then(policies => expect(policies).toEqual([])));

  it('should getPolicyByIdSid, by x509id, statement id', async () =>
    ngacRepo(context)
      .getPolicyByIdSid(x509id, sid)
      .then(({ sid }) => expect(sid).toBe('allowCreateDocument')));

  it('should fail getPolicyByIdSid: wrong x509id', async () =>
    ngacRepo(context)
      .getPolicyByIdSid('incorrect x509id', sid)
      .then(policy => expect(policy).toEqual([])));

  it('should fail getPolicyByIdSid: wrong sid', async () =>
    ngacRepo(context)
      .getPolicyByIdSid(x509id, 'incorrect sid')
      .then(policy => expect(policy).toEqual([])));

  it('should getMSPAttrByMSPID, by mspid', async () =>
    ngacRepo(context)
      .getMSPAttrByMSPID(mspid)
      .then(attributes => expect(attributes).toEqual([{ type: '1', key: 'mspid', value: 'Org1MSP' }])));

  it('should fail getMSPAttrByMSPID: wrong mspid', async () =>
    ngacRepo(context)
      .getMSPAttrByMSPID('incorrect mspid')
      .then(attributes => expect(attributes).toEqual([])));

  it('should getResourceAttrByURI', async () =>
    ngacRepo(context)
      .getResourceAttrByURI(uri)
      .then(attributes => expect(attributes[0].key).toEqual('createDocument')));

  it('should fail getResourceAttrByURI: wrong uri', async () =>
    ngacRepo(context)
      .getResourceAttrGroupByURI('incorrect-uri')
      .then(attributes => expect(attributes).toEqual([])));

  it('should addPolicy', async () => {
    const policy = createPolicy({
      context,
      policyClass: 'unit-test',
      sid: 'allowCreateTrade',
      allowedEvents: ['TradeCreated'],
      uri: `${NAMESPACE.MODEL}/Org0MSP/trade/trade_unit_01`
    });
    let policyAdded: Policy;
    await ngacRepo(context)
      .addPolicy(policy)
      .then(policy => {
        policyAdded = policy;
        expect(policy).toMatchSnapshot();
      });
    await ngacRepo(context)
      .getPolicyByIdSid(x509id, policy.sid)
      .then(({ sid }) => expect(sid).toEqual(policyAdded.sid));
  });

  it('should deletePolicyByIdSid', async () => {
    await ngacRepo(context)
      .deletePolicyByIdSid(x509id, 'allowCreateTrade')
      .then(keyDeleted => expect(keyDeleted).toBe(`"${x509id}""allowCreateTrade"`));
    await ngacRepo(context)
      .getPolicyByIdSid(x509id, 'allowCreateTrade')
      .then(policy => expect(policy).toEqual([]));
  });

  it('should deletePolicyById', async () => {
    // const id01 = 'id-deletePolicyById';
    context.clientIdentity.getMSPID.mockImplementation(() => 'Org0MSP');
    const policy1 = createPolicy({
      context,
      policyClass: 'unit-test',
      sid: 'allowCreateLoan',
      allowedEvents: ['LoanCreated'],
      uri: `${NAMESPACE.MODEL}/Org0MSP/loan/loan_unit_01`
    });
    const policy2 = createPolicy({
      context,
      policyClass: 'unit-test',
      sid: 'allowSignature',
      allowedEvents: ['Signed'],
      uri: `${NAMESPACE.MODEL}/Org0MSP/loan/loan_unit_01`
    });
    let policy1Added: Policy;
    let policy2Added: Policy;

    await ngacRepo(context)
      .addPolicy(policy1)
      .then(policy => (policy1Added = policy));

    await ngacRepo(context)
      .addPolicy(policy2)
      .then(policy => (policy2Added = policy));

    const id = createId(['Org0MSP', 'Admin@example.com']);
    await ngacRepo(context)
      .deletePolicyById(id)
      .then(keysDeleted => keysDeleted.forEach(key => expect(key.startsWith(`"${id}"`)).toBe(true)));
  });

  it('should addMSPAttr', async () => {
    const mspId = 'Org0MSP';
    const mspAttrs = [{ type: '1', key: 'mspid', value: 'Org0MSP' }];
    const mspResource = createMSPResource({ context, mspId, mspAttrs });
    let mspAttrAdded;
    await ngacRepo(context)
      .addMSPAttr(mspResource)
      .then(attrs => {
        mspAttrAdded = attrs;
        expect(attrs).toEqual(mspAttrs);
      });
    await ngacRepo(context)
      .getMSPAttrByMSPID('Org0MSP')
      .then(attrs => expect(attrs).toEqual(mspAttrs));
  });

  it('should deleteMSPAttrByMSPID', async () => {
    await ngacRepo(context)
      .deleteMSPAttrByMSPID('Org0MSP')
      .then(result => expect(result).toEqual('"Org0MSP"'));
    await ngacRepo(context)
      .getMSPAttrByMSPID('Org0MSP')
      .then(attrs => expect(attrs).toEqual([]));
  });

  it('should addResourceAttr', async () => {
    context.clientIdentity.getMSPID.mockImplementation(() => 'Org1MSP');
    const entityName = 'Trade';
    const entityId = 'unit_test_1010';
    const resourceAttrs = [{ type: '1', key: 'owner', value: 'me' }];
    const resource = createResource({
      context,
      entityId,
      entityName,
      resourceAttrs
    });
    await ngacRepo(context)
      .addResourceAttr(resource)
      .then(attrs => expect(attrs).toEqual(resourceAttrs));

    await ngacRepo(context)
      .getResourceAttrGroupByURI(resource.uri)
      .then(attrs => expect(attrs).toEqual(resourceAttrs));
  });

  it('should deleteReourceAttrByURI', async () => {
    const uri = `model/Org1MSP/Trade/unit_test_1010`;
    await ngacRepo(context)
      .deleteReourceAttrByURI(uri)
      .then(key => expect(key).toEqual('"model""Org1MSP""Trade""unit_test_1010"'));
    await ngacRepo(context)
      .getResourceAttrGroupByURI(uri)
      .then(attrs => expect(attrs).toEqual([]));
  });
});
