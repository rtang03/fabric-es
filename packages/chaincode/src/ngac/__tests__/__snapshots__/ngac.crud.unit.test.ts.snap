// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgacRepo CRUD Tests should addPolicy 1`] = `
Object {
  "allowedEvents": Array [
    "TradeCreated",
  ],
  "attributes": Object {
    "uri": "model/Org0MSP/trade/trade_unit_01",
  },
  "effect": "Allow",
  "key": "\\"Org1MSP::Admin@example.com\\"~\\"allowCreateTrade\\"",
  "policyClass": "unit-test",
  "sid": "allowCreateTrade",
}
`;
